presto-hive/src/main/java/com/facebook/presto/hive/util/AsyncQueue.java

```diff
@@ -13,103 +13,111 @@
+        CompletableFuture<?> offerResubmissionFuture = offerResubmissionFutureReference.get();
+
+        int newSize = queueSize.incrementAndGet();
+        elements.add(element);
+        if (newSize == 1) {
+            signalFetchResubmissionAndReset();
         }
-        queue.add(element);
-        notEmptyFuture.get().complete(null);
+        return newSize < softMaxQueueSize ? COMPLETED_FUTURE : offerResubmissionFuture;
     }
 
-    /**
-     * Gets a batch of elements from the queue.  If queue is not empty, a
-     * completed future containing the queued elements is returned.  If the
-     * queue is empty, an incomplete future is returned that completes when
-     * an element is added.
-     * <p>
-     * It is possible that an empty list will be returned as another reader
-     * may have consumed all of the data.  A caller should read until, the
-     * {@code isFinished} method returns {@code true}.
-     *
-     * @param maxSize the maximum number of elements to return.
-     */
-    public CompletableFuture<List<T>> getBatchAsync(int maxSize)
+    private int drainTo(Collection<T> collection, int maxSize)
     {
-        return notEmptyFuture.get().thenApply(x -> getBatch(maxSize));
+        int drained = elements.drainTo(collection, maxSize);
+        if (drained > 0) {
+            reduceQueueCount(drained);
+        }
+        return drained;
     }
 
     private List<T> getBatch(int maxSize)
     {
-        // take up to maxSize elements from the queue
-        List<Object> elements = new ArrayList<>(maxSize);
-        queue.drainTo(elements, maxSize);
-
-        // if the queue is empty and the current future is finished, create a new one so
-        // a new readers can be notified when the queue has elements to read
-        if (queue.isEmpty() && !finished.get()) {
-            CompletableFuture<?> future = notEmptyFuture.get();
-            if (future.isDone()) {
-                notEmptyFuture.compareAndSet(future, new CompletableFuture<>());
-            }
+        List<T> list = new ArrayList<>(maxSize);
+        drainTo(list, maxSize);
+        return list;
+    }
+
+    public CompletableFuture<List<T>> getBatchAsync(int maxSize)
+    {
+        // Grab the future first.
+        CompletableFuture<?> future = fetchResubmissionFutureReference.get();
+
+        List<T> list = getBatch(maxSize);
+        if (list.isEmpty()) {
+            return future.thenApply(x -> getBatch(maxSize));
         }
+        else {
+            return CompletableFuture.completedFuture(list);
+        }
+    }
 
-        // if someone added an element while we were updating the state, complete the current future
-        if (!queue.isEmpty() || finished.get()) {
-            notEmptyFuture.get().complete(null);
+    private void reduceQueueCount(int reduceBy)
+    {
+        if (reduceBy > 0) {
+            int newSize = queueSize.addAndGet(-reduceBy);
+            Verify.verify(newSize >= 0, "INVARIANT");
```

Comment:
any particular reason to use this instead of `checkState`? `checkState` seems to be what we use in most other parts of the code base